README.md
# GovUkDateFields
## Day/Month/Year input boxes for Rails projects

## IMPORTANT

There are 2 versions of this gem you can choose.

* Version `3.1.0` or lower will produce HTML markup and styles compatible with the old version of the GDS design system.

* Version `4.0.0` or greater will produce HTML markup and styles compatible with the new version of the [GDS design system](https://design-system.service.gov.uk/components/date-input/).  
In the future, the functionality of this gem will be probably merged into a gem that will contain other common form components, 
like radios, text areas, checkboxes, etc.

## Overview

The GOV.UK standard for [date fields](https://govuk-elements.herokuapp.com/form-elements/example-date/) on 
web forms is to use three boxes - one each for day, month and year, rather 
than the drop_down boxes that comes as standard on rails projects. This is because research has shown that 
[some users find it difficult to use dropdown lists](https://paper.dropbox.com/doc/Date-picker-vPs1AAOz3bXLm7P3rCrJm) 
and scrolling through a long list of years (eg for date of birth) can be slower than entering 4 digits using the keyboard.

This gem provides the methods required to 
easily display and validate dates entered into forms in this way.


## Getting Started

Follow the four simple steps below.  In the example below, we assume there is an Employee model with 
attributes, ```name```, ```dob```, ```joined```, the last two attributes being ```gov_uk_dates```.  Note 
that these must be defined on the database as ```date``` fields, not ```datetime```.

### 1. Add the Gem to your Gemfile

Add the following line to your Gemfile:

    gem 'gov_uk_date_fields'

Then run bundle install.  (Note: version 2+ requires Rails 5+, see [gov_uk_date_fields](https://rubygems.org/gems/https://rubygems.org/gems/gov_uk_date_fields/versions) on Rubygems for previous versions.)  
Choose version `3.X` for old design system, and `>= 4.X` for new design system.

### 2. Get your app to load the GovUkDateFields assets (only for versions < 4.X of the gem)

#### 2.1 Stylesheets

Add the following line to the file ```app/assets/stylesheets/application.css.scss```

    @import 'govuk_date_fields';
    
#### 2.2 Javascript

Add the following line to the file ```app/assets/javascripts/application.js```

    //= require govuk_date_fields

Note: the javascript is optional (required only for the 'Today' button), and requires jQuery.

### 3. Tell your models attributes are gov_uk_dates

To specify date fields on the database that are to be rendered and validated as gov_uk_date_fields, 
simply add an acts_as_gov_uk_date to your model wth the attribute names of the date fields, e.g.:

    class Employee < ActiveRecord::Base
      acts_as_gov_uk_date :dob, :joined
    end

This tells the gem that there are two columns of type date on the database record for this model, ```dob``` and ```joined```, which will be rendered as three boxes on the form.

#### 3.1 Options that can be passed to ```acts_as_gov_uk_date```

 * ```:validate_if``` Determines whether or not to validate the date as part of the model's validation routines.  Pass in the name of a method which returns true or false. If
   not specified, true is assumed.

 * ```:error_clash_behaviour``` - Determines what action to take with multiple error messages if the model's validation has added an error message to the errors hash for this field, and this gem also determines
   that it is invalid.  Possible values are:
   *  :append_gov_uk_date_field_error - This gem's 'Invalid date' message will be added to the errors hash, resulting in both error messages being displayed
   *  :omit_gov_uk_date_field_error - The gem's 'Invalid date' message is discarded - ony the model's validation error message will be displayed
   *  :override_with_gov_uk_date_field_error - The model's error message is discarded and this gem's 'Invalid date' message added to the error hash.  Only the invalid date will be displayed.

   If the ```:error_clash_behaviour``` option is not specified, ```:append_gov_uk_date_field_error``` is assumed.

#### 3.2 Examples

    acts_as_gov_uk_date :date_of_birth, :date_joined, 
                        validate_if: :perform_validation?, 
                        error_clash_behaviour: :override_with_gov_uk_date_field_error

    def perform_validation?
      return true unless self.draft?
    end


You can also specify a Proc or a Symbol pointing to a method that checks whether the date validations should be performed. This is optional and by default validations are always performed. Example:

    acts_as_gov_uk_date :dob, :joined, validate_if: :perform_validation?

### 4. Permit the date parameters in your controller

Your controller needs to permit three parameters for each gov_uk_date field, so in the example above the 
employee_params method of the EmployeesController would be:

    def employee_params
      params.require(:employee).permit(:name, :dob_dd, :dob_mm, :dob_yyyy, :joined_dd, :joined_mm, :joined_yyyy)
    end


### 5. Update your form to render the three boxes

Use the gov_uk_date_field method that this gem adds into FormBuilder to create the three
date field boxes in the form:

    <%= form_for(@employee) do |f| %>
       <%= f.gov_uk_date_field :dob, legend_text: 'Date of birth' %>
   
       <%= f.gov_uk_date_field :joined, legend_text: 'Date joined' %>
    <% end %>

#### 5.1. Options passed to gov_uk_date_field (versions < 4.x of the gem)

The FormBuilder method gov_uk_date_field takes two parameters:
  
  * the method on the model that the FormBuilder is encapsulating
  
  * an option hash describing how the date fields should be rendered:
  
    - **an empty hash or nil**: means the date fields will be rendered as per version 0.1.0 of this gem, 
      that is just three input fields with no encapsulating fieldset, divs, or legends.  This is 
      now deprecated and should no longer be used, but is included for backward 
      compatibility with versions 0.0.1 and 0.0.2.
      
    - **placeholder**: see below for an explanation of how to specify placeholders.  This is now deprecated 
      and should no longer be used, but is included for backward compatibility with versions 0.0.1 and 0.0.2.
      
    - **legend_text**: The text that is to be used as the title of the Date field set.
    
    - **legend_class**: The CSS class that is to be used for the legend
    
    - **form_hint_text**: The text that is to advise the user how to fill in the form.  If not specified, 
      the text "For example, 31 3 1980" will be used.
      
    - **id**: The id to be given to the fieldset.  If absent, it will be autogenerated.  This setting also affects the id of the hint element:
      if specified, then the hint element will have the fieldset id appended by '-hint', otherwise it will have the autogenerated id appended by '-hint'.  The hint-id is
      referred to by the aria-described-by attribute on the input element.

    - **error_messages**: Error messages to be attached to the field, if not the string in the errors collection of the object.
      This is useful if the error messages are held in a translation file for example - the client should fetch the translations and
      pass in as an array of strings as the value for this option.
      
    - **today_button**: If present, a "Today" button which, when presssed, will populate the form with today's date
      will be generated.  The value of this option should either be:
       * true: a Today button with the default value of "button" will be generated
       * {class: "button-class"} a Today button with the specified CSS class(es) will be generated
       * false: a Today button will not be generated

##### 5.1.1 Placeholders:

  Supply your own

    <%= f.gov_uk_date_field :dob, placeholders: { day: 'dd', month: 'mm', year: 'yyyy' } %>

  or enable defaults (DD/MM/YYYY)

    <%= f.gov_uk_date_field :dob, placeholders: true %>

      
#### 5.2. Options passed to gov_uk_date_field (versions >= 4.x of the gem)

* **i18n_attribute**: optionally, this will be the locales lookup key in hint and legend. The "real" attribute will still take precedence in all other places or if the `i18n_attribute` is not found in the locales.

* **legend_options**: it can contain any of the following:
  - `class`: will append any given class or classes to those generated by the gem.
  
  - `page_heading`: can be true or false. Default is true. Refer to https://design-system.service.gov.uk/get-started/labels-legends-headings/ for more details.
  
  - `visually_hidden`: can be true or false. Default is false.

All the strings will be picked from i18n locales, including legend, hint if any, and errors.

Example:

    <%= f.gov_uk_date_field :dob, i18n_attribute: :applicant_dob,
        legend_options: { page_heading: false, visually_hidden: true } %>

### 6.  You're done!

You're ready to go.

If the user enters values into the day/month/year boxes that can't be turned into a date, the attribute will be marked as 
invalid in the model's errors hash during the validation cycle, and the entered values will be displayed back to
the user on the form.


## Developing

### Testing

Prepare the test database (run this in the `test/dummy/` directory)

    bundle exec rake db:setup

Run tests either with the rake task:

    bundle exec rake test

Or run test files directly from the command-line:

    bundle exec ruby -I. test/dummy/test/models/form_date_test.rb

    # You can run with on a specific test by name
    bundle exec ruby -I. test/dummy/test/models/employee_test.rb -n test_creating_a_new_employee_with_invalid_dates_is_invalid


## FAQs

### What is the HTML that this gem produces?

Given an employee object with the dob attribute set to
  
    Date.new(1965, 7, 13)
  
the following call to the gov_uk_date_field method:

    f.gov_uk_date_field(:dob, legend_text: 'Date of birth', legend_class: 'govuk-legend')
    
will produce the following HTML with gem version < 3.0.0:

      <fieldset>
        <legend class="govuk_legend">Date of birth</legend>
        <div class="form-date">
          <p class="form-hint" id="dob-hint">For example, 31 3 1980</p>
          <div class="form-group form-group-day">
            <label for="dob-day">Day</label>
            <input class="form-control" 
                   id="dob-day" 
                   name="dob-day" 
                   type="number" 
                   pattern="[0-9]*" 
                   min="0"
                   max="31" 
                   aria-describedby="dob-hint" 
                   value="13">
          </div>
          <div class="form-group form-group-month">
            <label for="dob-month">Month</label>
            <input class="form-control" 
                   id="dob-month" 
                   name="dob-month" 
                   type="number" 
                   pattern="[0-9]*" 
                   min="0" 
                   max="12" 
                   value="7">
          </div>
          <div class="form-group form-group-year">
            <label for="dob-year">Year</label>
            <input class="form-control" 
                   id="dob-year" 
                   name="dob-year" 
                   type="number" 
                   pattern="[0-9]*" 
                   min="0" 
                   max="2016" 
                   value="1965">
          </div>
        </div>
      </fieldset>

the following HTML with gem version >= 3.0.0, < 3.1.0:

    <div class="form-group gov_uk_date">
        <fieldset>
            <legend>
                <span class="form-label-bold">Date of birth</span>
                <span class="form-hint" id="dob-hint">For example, 31 3 1980</span>
            </legend>
            <div class="form-date">
                <div class="form-group form-group-day">
                    <label for="employee_dob_dd">Day</label>
                    <input class="form-control" 
                           id="employee_dob_dd" 
                           name="employee[dob_dd]" 
                           type="number" 
                           min="0" 
                           max="31" 
                           aria-describedby="dob-hint" 
                           value="13">
                </div>

                <div class="form-group form-group-month">
                    <label for="employee_dob_mm">Month</label>
                    <input class="form-control" 
                           id="employee_dob_mm" 
                           name="employee[dob_mm]" 
                           type="number" 
                           min="0" 
                           max="12" 
                           value="7">
                </div>

                <div class="form-group form-group-year">
                    <label for="employee_dob_dd_yyyy">Year</label>
                    <input class="form-control" 
                           id="employee_dob_yyy" 
                           name="employee[dob_yyy]" 
                           type="number" 
                           min="0" 
                           max="2100" 
                           value="1965">
                </div>
            </div>
        </fieldset>
    </div>

the following HTML with gem version >= 3.1.0 and < 4.0.0:

    <div class="form-group gov_uk_date" id="employee_dob">
        <fieldset>
            <legend>
                <span class="form-label-bold">Date of birth</span>
                <span class="form-hint" id="employee_dob-hint">For example, 31 3 1980</span>
            </legend>
            <div class="form-date">
                <div class="form-group form-group-day">
                    <label for="employee_dob_dd">Day</label>
                    <input class="form-control" 
                           id="employee_dob_dd" 
                           name="employee[dob_dd]" 
                           type="number" 
                           min="0" 
                           max="31" 
                           aria-describedby="employee_dob-hint" 
                           value="13">
                </div>

                <div class="form-group form-group-month">
                    <label for="employee_dob_mm">Month</label>
                    <input class="form-control" 
                           id="employee_dob_mm" 
                           name="employee[dob_mm]" 
                           type="number" 
                           min="0" 
                           max="12" 
                           value="7">
                </div>

                <div class="form-group form-group-year">
                    <label for="employee_dob_yyy">Year</label>
                    <input class="form-control" 
                           id="employee_dob_yyy" 
                           name="employee[dob_yyy]" 
                           type="number" 
                           min="0" 
                           max="2100" 
                           value="1965">
                </div>
            </div>
        </fieldset>
    </div>

and the following HTML with gem version >= 4.0.0: [html fixture](test/dummy/test/fixtures/files/date_fields_no_error.html)

See other examples in the file `test/form_fields_test.rb`

Please note, version 4.0.0 and 4.0.1 will produce inputs with `type="number"` whereas versions >= 4.1.0 will produce inputs with `type="text" inputmode="numeric"`. These are semantically correct and provides a better experience in most browsers, including mobile.

## Licencing

This project rocks and uses MIT-LICENSE.


CHANGELOG.md
# 4.2.0

* BREAKING CHANGE: drop support for [End of life rails versions](https://endoflife.date/rails), name less that Rails 5.2

* Fixes deprecation warnings when using rails 6.1, which may break in rails 6.2.

  ```
  DEPRECATION WARNING: Calling `<<` to an ActiveModel::Errors message array in order to add an error is deprecated.
  Please call `ActiveModel::Errors#add` instead.
  ```
  ```
  DEPRECATION WARNING: Calling `clear` to an ActiveModel::Errors message array in order to delete all errors is deprecated.
  Please call `ActiveModel::Errors#delete` instead.
  ```
  ```
  DEPRECATION WARNING: ActiveModel::Errors#keys is deprecated and will be removed in Rails 6.2.
  To achieve the same use:
    errors.attribute_names
  ```

# 4.1.0

* Markup will produce inputs with `type="text" inputmode="numeric"`.

These are semantically correct and provides a better experience in most browsers, including mobile. More details in [this thread](https://github.com/alphagov/govuk-frontend/issues/1449#issuecomment-504006087)

This change **might** break your code if you target the input (in javascript, CSS or cucumber for example) by `type="number"`.

# 4.0.1

* Allow alternative i18n locales. [Read more details](https://github.com/ministryofjustice/gov_uk_date_fields/pull/36)

# 4.0.0

* **BREAKING CHANGE**: Support for new design system
* The gem will produce markup and style classes in line with the new [design system](https://design-system.service.gov.uk/components/date-input/)
* The gem will pick all the strings from i18n locales, and the ability to pass hardcoded strings has been removed
* There is no longer an option to show a 'today' button, and thus all related javascript has been removed
* Removed the stylesheet file, as all the styling is now coming from the design system

# Older versions

Please refer to the [releases page](https://github.com/ministryofjustice/gov_uk_date_fields/releases) and the [commits history](https://github.com/ministryofjustice/gov_uk_date_fields/commits/master).


test/test_helper.rb
# Configure Rails Environment
ENV["RAILS_ENV"] = "test"

require File.expand_path("../../test/dummy/config/environment.rb",  __FILE__)
ActiveRecord::Migrator.migrations_paths = [File.expand_path("../../test/dummy/db/migrate", __FILE__)]
require "rails/test_help"
require "minitest/reporters"
require 'rails-controller-testing'

Rails::Controller::Testing.install

# configure minitest reporting
reporter_options = { color: true }
Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new(reporter_options)

# Filter out Minitest backtrace while allowing backtrace from other libraries
# to be shown.
Minitest.backtrace_filter = Minitest::BacktraceFilter.new


# Load support files
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each { |f| require f }

# Load fixtures from the engine
if ActiveSupport::TestCase.respond_to?(:fixture_path=)
  ActiveSupport::TestCase.fixture_path = File.expand_path("../fixtures", __FILE__)
  ActionDispatch::IntegrationTest.fixture_path = ActiveSupport::TestCase.fixture_path
  ActiveSupport::TestCase.fixtures :all
end



def assert_false(actual_value, message = nil)
  assert_equal false, actual_value, "#{message} was expected to be false, is #{actual_value}"
end


def assert_true(actual_value, message = nil)
  assert_equal true, actual_value, "#{message} was expected to be true, is #{actual_value}"
end


def assert_html_equal(actual_html, expected_html)
  assert_equal(squash_html(actual_html), squash_html(expected_html), "\nExpected: #{squash_html(expected_html)}\nGot:      #{squash_html(actual_html)}")
end

def squash_html(html)
  html.gsub("\n", "").gsub(/\s*</, "<").gsub(/>\s*/, ">")
end





test/dummy/db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20150901103114) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "employees", force: :cascade do |t|
    t.string   "name"
    t.date     "dob"
    t.date     "joined"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

end


lib/gov_uk_date_fields/form_date.rb
module GovUkDateFields

  class FormDateError < RuntimeError; end


  class FormDate

    attr_accessor :dd, :mm, :yyyy, :date, :temp
    
    # This method cannot be called directly: use .new_from_date or .new_from_params
    def initialize(dd, mm, yyyy, date = nil)
     @dd            = dd
     @mm            = mm
     @yyyy          = yyyy
     @date          = date
     @valid         = date.nil? ? validate_date : true
    end
  
    # Creates a FormDate object from the date, and writes it to @_<attr_name>
    # on the object, and updates the attr[<attr_name>] with the real date object
    #
    def self.set_from_date(object, attr_name, date)
      if date.nil?
        form_date = new('', '', '')
      else
        form_date = new(date.day.to_s, date.month.to_s, date.year.to_s, date)
      end
      object.instance_variable_set("@_#{attr_name}", form_date)
    end



    def self.set_date_part(date_part, object, attr_name, value)
      form_date = object.instance_variable_get("@_#{attr_name}".to_sym) || FormDate.nil_date
      form_date.send("#{date_part}=", value)
      form_date.create_date_from_date_parts
      object.__send__("#{attr_name}=", form_date.date)
      object.instance_variable_set("@_#{attr_name}", form_date)
    end

    def self.nil_date
      new('', '', '')
    end


    def create_date_from_date_parts
      @valid = true
      if @dd.blank? && @mm.blank? && @yyyy.blank?
        @date = nil
      else
        months = %w{ jan feb mar apr may jun jul aug sep oct nov dec }
        mm_as_int = months.include?(@mm.downcase) ? months.index(@mm.downcase) + 1 : @mm.to_i
        begin
          @date = Date.new(@yyyy.to_i, mm_as_int, @dd.to_i)
        rescue ArgumentError => err
          raise err unless err.message == 'invalid date'
          @valid = false
        end
      end
    end


    # Creates a FormDate object from the params, and writes it to the instance variable @_<attr_name> 
    # on the object, and updates the attr[attr_name] with the real date object
    #
    # If the dob_dd, dob_mm, dob_yy are not in the params hash, or if they are all blank, the object is 
    # not updated.
    #
    # def self.set_from_params(object, attr_name, params)
    #   params = HashWithIndifferentAccess.new(params) unless params.is_a?(HashWithIndifferentAccess)
    #   return if attr_not_present_in_params?(attr_name, params)
    #   form_date = new(params["#{attr_name}_dd"], params["#{attr_name}_mm"], params["#{attr_name}_yyyy"] )
    #   update_object(object, attr_name, form_date)
    # end

    def valid?
      @valid
    end

    private_class_method :new

    private 


    def transfer_temp_fields_to_live
      @dd = @temp[:dd]
      @mm = @temp[:mm]
      @yyyy = @temp[:yyyy]
    end

    def self.attr_not_present_in_params?(attr_name, params)
      attr_missing_in_params?(attr_name, params)  || all_blanks_in_params?(attr_name, params)
    end

    def self.attr_missing_in_params?(attr_name, params) 
      params["#{attr_name}_dd"].nil? &&  params["#{attr_name}_mm"].nil? &&  params["#{attr_name}_yyyy"].nil?
    end

    def self.all_blanks_in_params?(attr_name, params)
      params["#{attr_name}_dd"].blank? &&  params["#{attr_name}_mm"].blank? &&  params["#{attr_name}_yyyy"].blank?
    end


    def validate_date
      return true if @dd.blank? && @mm.blank? && @yyyy.blank?
      months = %w{ jan feb mar apr may jun jul aug sep oct nov dec }

      mm_as_int = months.include?(@mm.downcase) ? months.index(@mm.downcase) + 1 : @mm.to_i
      begin
        @date = Date.new(@yyyy.to_i, mm_as_int, @dd.to_i)
      rescue ArgumentError => err
        return false if err.message == 'invalid date'
        raise err
      end
    end

  end

end


lib/gov_uk_date_fields/form_fields.rb
module GovUkDateFields

  class FormFields
    VALID_OPTIONS = [:legend_text, :legend_class, :form_hint_text, :id, :placeholders, :error_messages, :today_button]

    DATE_SEGMENTS = {
      day:    '_dd',
      month:  '_mm',
      year:   '_yyyy',
    }

    DEFAULT_PLACEHOLDERS = {
      day: 'DD',
      month: 'MM',
      year: 'YYYY'
    }

    def initialize(form, object_name, attribute, options={})
      @form               = form
      @object             = form.object
      @object_name        = object_name
      @attribute          = attribute
      @options            = options
      @day_value          = @object.send("#{@attribute}_dd")&.gsub(/\D/, '')
      @month_value        = @object.send("#{@attribute}_mm")&.gsub(/\D/, '')
      @year_value         = @object.send("#{@attribute}_yyyy")&.gsub(/\D/, '')
      @form_hint_text     = @options[:form_hint_text] || "For example, 31 3 1980"
      @fieldset_required  = false
      @fieldset_id        = @options[:id]
      @error_messages     = @options[:error_messages]
      @hint_id            = form_group_hint_id
      @today_button       = @options[:today_button] || false
      parse_options
    end

    def raw_output
      if fieldset_required?
        generate_input_fields
      else
        generate_old_style_input_fields
      end
    end

    def output
      raw_output.html_safe
    end

  private

    def error_for_attr?
      error_attribute_names.include?(@attribute) && @object.errors[@attribute].any?
    end

    def error_attribute_names
      if @object.errors.respond_to?(:attribute_names)
        @object.errors.attribute_names
      else
        @object.errors.keys
      end
    end

    def fieldset_required?
      @fieldset_required
    end

    def generate_old_style_input_fields
      %Q[
        #{@form.text_field(@attribute, field_options(@day_value, html_id(:day), html_name(:day), placeholder(:day), 2))}
        #{@form.text_field(@attribute, field_options(@month_value, html_id(:month), html_name(:month), placeholder(:month), 3))}
        #{@form.text_field(@attribute, field_options(@year_value, html_id(:year), html_name(:year), placeholder(:year), 4))}
      ]
    end

    def generate_start_fieldset
      %Q|
        #{generate_fieldset_tag}
          #{generate_legend_tag}
            <span class="form-label-bold">#{@options[:legend_text]}</span>
            <span class="form-hint" id="#{@hint_id}">#{@form_hint_text}</span>
            #{generate_error_message}
          </legend>
          <div class="form-date">
      |
    end

    def generate_fieldset_tag
      css_class = "form-group gov_uk_date"
      css_class += " form-group-error" if error_for_attr?

      result = %Q|
                <div class="#{css_class}"|
      result += %Q| id="#{form_group_id}"|
      result += ">"
      result += %Q| <fieldset|
      result += ">"
      result
    end

    def generate_end_fieldset
      "</div></fieldset></div>"
    end

    def generate_legend_tag
      if @options.key?(:legend_class)
        %Q|<legend class="#{@options[:legend_class]}">|
      else
        "<legend>"
      end
    end

    def generate_error_message
      result = ''
      if error_for_attr?
        result = "<ul>"
        if @error_messages.nil?
          @error_messages = @object.errors[@attribute]
        end
        @error_messages.each do |message|
          result += %Q|<li><span class="error-message">#{message}</span></li>|
        end
        result += "</ul>"
      end
      result
    end

    def generate_input_fields
      result = generate_start_fieldset
      result += generate_today_button if @today_button
      result += generate_day_input_field(@day_value) + generate_month_input_field(@month_value) + generate_year_input_field(@year_value)
      result += generate_end_fieldset
      result
    end

    def today_button_class
      if @today_button.is_a?(Hash) && @today_button.key?(:class)
        @today_button[:class]
      else
        'button'
      end
    end

    def generate_today_button
      %Q|<a class="#{today_button_class}" role="button" href="#">Today</a>|
    end

    def generate_start_div
      %Q|<div class="form-group form-group-day">|
    end

    def generate_end_div
      %Q|</div>|
    end

    def generate_day_input_field(day_value)
      css_class = "form-control"
      css_class += " form-control-error" if error_for_attr?

      result = %Q|
          <div class="form-group form-group-day">
            <label for="#{html_id(:day)}">Day</label>
            <input class="#{css_class}" id="#{html_id(:day)}" name="#{html_name(:day)}" type="number" min="0" max="31" aria-describedby="#{@hint_id}" value="#{day_value}">
          </div>
      |
    end

    def generate_month_input_field(month_value)
      css_class = "form-control"
      css_class += " form-control-error" if error_for_attr?

      result = %Q|
        <div class="form-group form-group-month">
          <label for="#{html_id(:month)}">Month</label>
          <input class="#{css_class}" id="#{html_id(:month)}" name="#{html_name(:month)}" type="number" min="0" max="12" value="#{month_value}">
        </div>
      |
    end

    def generate_year_input_field(year_value)
      css_class = "form-control"
      css_class += " form-control-error" if error_for_attr?

      result = %Q|
        <div class="form-group form-group-year">
          <label for="#{html_id(:year)}">Year</label>
          <input class="#{css_class}" id="#{html_id(:year)}" name="#{html_name(:year)}" type="number" min="0" max="2100" value="#{year_value}">
        </div>
      |
    end

    def parse_options
      validate_option_keys
      if @options.key?(:legend_text) || @options.key?(:id)
        @fieldset_required = true
      else
        if @options.key?(:legend_class) || @options.key?(:form_hint_text)
          raise ArgumentError.new("Invalid combination of options: You must specifigy :legend_text if :legend_class or :form_hint_text are specified")
        end
      end
    end

    def validate_option_keys
      @options.keys.each do |option_key|
        unless VALID_OPTIONS.include?(option_key)
          raise ArgumentError.new("Invalid option key: #{option_key.inspect}")
        end
      end
    end


    def placeholder(part)
      if @options[:placeholders] == true
        DEFAULT_PLACEHOLDERS[part]
      else
        @options[:placeholders][part] || DEFAULT_PLACEHOLDERS[part] if @options[:placeholders]
      end
    end

    def html_id(date_segment)
      brackets2underscore(html_name(date_segment))
    end

    def html_name(date_segment)
      "#{@object_name}[#{@attribute}#{DATE_SEGMENTS[date_segment]}]"
    end

    def brackets2underscore(string)
      string.tr('[','_').tr(']', '_').gsub('__', '_').gsub(/_$/, '')
    end

    def attribute_prefix
      brackets2underscore(@object_name.to_s)
    end

    def form_group_id
      # If an `id` option was passed, this takes precedence, to ensure backward compatibility
      return @fieldset_id if @fieldset_id

      group_id = [attribute_prefix, @attribute]
      group_id.unshift('error') if error_for_attr?
      group_id.join('_')
    end

    def form_group_hint_id
      # If an `id` option was passed, this takes precedence, to ensure backward compatibility
      return "#{@fieldset_id}-hint" if @fieldset_id

      [attribute_prefix, @attribute].join('_') + '-hint'
    end

    def field_options(value, id, name, placeholder, size)
      {
        value: value,
        id: id,
        name: name,
        placeholder: placeholder,
        size: size
      }
    end
  end
end
